### PR TITLE
Update documentation of Trace

### PR DESCRIPTION
**Pull request by eb8680**
Part of #826 

This PR updates the docstrings of `Trace` with more information and some usage examples.

### CODE REVIEW

File: pyro/poutine/trace.py [hunk 1]
```diff
@@ -32,7 +35,54 @@ def fresh_copy(self):
...
+
+    Each node (or site) in a trace contains the name, input and output value of the site,
+    as well as additional metadata added by inference algorithms or user annotation.
+    In the case of ``pyro.sample``, the trace also includes the stochastic function
+    at the site, and any observed data added by users.
+
+    Consider the following Pyro program:
+
+        >>> def model(x):
+        ...     s = pyro.param("s", torch.tensor([0.5]))
```
Comment (fritzo) [inline pyro/poutine/trace.py]: nit: The scalar value below suggests this should be `torch.tensor(0.5)`?

File: pyro/poutine/trace.py [hunk 2]
```diff
@@ -50,49 +100,84 @@ def __init__(self, *args, **kwargs):
...
+        """
+        Work around cyclic reference bugs in ``networkx.DiGraph``
+        See ``https://github.com/uber/pyro/issues/798``
+        """
         self._graph.__dict__.clear()
 
     @property
     def nodes(self):
+        """
+        Identical to ``networkx.DiGraph.nodes``
```
Comment (fritzo) [inline pyro/poutine/trace.py]: If you link to sphinx docs in our conf.py
```diff
diff --git a/docs/source/conf.py b/docs/source/conf.py
index 73da0c17..65175989 100644
--- a/docs/source/conf.py
+++ b/docs/source/conf.py
@@ -175,6 +175,7 @@ texinfo_documents = [
 intersphinx_mapping = {
     'python': ('https://docs.python.org/3/', None),
     'torch': ('http://pytorch.org/docs/master/', None),
+    'networkx': ('https://networkx.github.io/documentation/stable/', None),
 }
 
 # document class constructors (__init__ methods):
```
then you could promote these to links via
```
"""
Identical to :attr:`networkx.DiGraph.nodes`
"""
```
(I think `:attr:` is correct here, others may need `:meth:`)

File: pyro/poutine/trace.py [hunk 3]
```diff
@@ -32,7 +35,54 @@ def fresh_copy(self):
 
 class Trace(object):
     """
-    Execution trace data structure
+    Execution trace data structure built on top of ``networkx.DiGraph``.
```
Comment (fritzo) [inline pyro/poutine/trace.py]: nit: promote to a link?
```
:class:`networkx.DiGraph`
```

File: pyro/poutine/trace.py [hunk 4]
```diff
@@ -194,6 +281,14 @@ def observation_nodes(self):
                 if node["type"] == "sample" and
                 node["is_observed"]]
 
+    @property
+    def param_nodes(self):
+        """
+        Gets a list of names of param sites
```
Comment (jpchen) [inline pyro/poutine/trace.py]: can you write this more sphinx-formally ie `:return: list of names of param sites`

**Pull request merged**